### PR TITLE
Feature/docusaurus v2 v3 update

### LINKED ISSUE #756: Migrate Docusaurus to version 3.0
**Is your feature request related to a problem? Please describe.**
Docusaurus 3.0 was released on 31st October 2023. Docusaurus 2.X entered maintenance mode and it is recommended to migrate to 3.0 until 31st January 2024.
We should therefore upgrade our documentation page until that date.

**Describe the solution you'd like**
The major release contains breaking changes. See official documentation for details: https://docusaurus.io/blog/releases/3.0

### PR DESCRIPTION
**Pull request by eliasarnold**
This PR includes an up-to-date version of the docusaurus documentation. All in all, the update was no big deal. The major change from version 2 to version 3 is the new and stricter Markdown parser (MDX). That is why some characters had to be escaped and some tags are entirely new (in Markdown).

### There are still some open questions:
1. In v3, there are more themes available, and their naming is different than in v2. The two themes, between which can be toggled (when clicking the "sun" icon on the SDLB website) can be changed in `docusaurus.config.js`
2. ~~To make the schema viewer appear in full screen, I had to set `<div style{{height: 1000}}` in `src\pages\json-schema-viewer.js`, which is not a good solution because the height is hardcoded. The attempt to set the height to `100%` or `100vh` was made with no positive outcome.~~

### CODE REVIEW

**Comment by pgruetter** [comment]
Hi @eliasarnold 
The target branch is wrong, please switch target branch to documentation.

File: src/pages/json-schema-viewer.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ function JsonSchemaViewer() {
 
     return (
         <Layout title="JsonSchemaViewer" noFooter={true} wrapperClassName="schema-viewer-wrapper">
-            <div style={{flex: 1}}>
+            <div style={{height: 1000}}>
```
Comment (eliasarnold) [inline src/pages/json-schema-viewer.js]: Does anyone know how this can be rewritten more generally? 'height: 100%' does not work. However, the current solution looks fine in the UI.

Comment (eliasarnold) [inline src/pages/json-schema-viewer.js]:
```suggestion
            <div style={{height: "100vh"}}>
```

File: static/json-schema-viewer/schemas/index.json
Comment (zzeekk) [inline static/json-schema-viewer/schemas/index.json:1]: This should not be committed, it should be generated by create-schema-index.js in github action.

File: yarn.lock
Comment (zzeekk) [inline yarn.lock:1]: We didn't commit the lock file before, why should we now?

File: create-schema-index.js
Comment (zzeekk) [inline create-schema-index.js:1]: Why is it deleted? It should be called during the Build.
Once this was a Docusaurus PostBuild Plugin: https://github.com/smart-data-lake/smart-data-lake/commit/2e5126cd7a2b9860bf6e3c3d990d3cb7accc2169.
Seems the plugin got lost somehow... and the script isn't called currently.
I wonder how this works for the current deployment...

**Review by zzeekk — COMMENTED** [review]
New docusaurus works well, had to upgrade node to version 18 for testing.
I further improved the schema viewer sizing and some other minor improvements.

I have added some points to discuss as well.

File: yarn.lock
Comment (eliasarnold) [reply yarn.lock:1]: Ah, now I remember: I started the migration process with a default docusaurus3 project which is created by running `npx create-docusaurus@latest my-website classic`. I accidentally committed the `.gitignore` file from this project 😂 I think this is also the reason for the missing PostBuild-Plugin (your next question).

File: blog/2022-04-06-SDL-snowpark/2022-04-06-SDL-snowpark.md [hunk 2]
```diff
@@ -12,7 +12,7 @@ hide_table_of_contents: false
 
 This article shows how to create one unified data pipeline that uses Spark to ingest data into Snowflake, and Snowpark to transform data inside Snowflake.
 
-<!--truncate-->
+\<!--truncate-->
```
Comment (pgruetter) [inline blog/2022-04-06-SDL-snowpark/2022-04-06-SDL-snowpark.md]: All these escapes to not work for me, see this one for example:
image

Same goes for all escapes further down. Why were they inserted?

File: docs/getting-started/part-1/select-columns.md [hunk 3]
```diff
@@ -234,7 +235,7 @@ podman run --rm -v ${PWD}/data:/mnt/data -v ${PWD}/target:/mnt/lib -v ${PWD}/con
 </TabItem>
 </Tabs>
 
-:::warning
+:::danger
```
Comment (pgruetter) [inline docs/getting-started/part-1/select-columns.md]: Why this change? Warning is still supported.

File: src/pages/json-schema-viewer.js (hunk 1 above)
Comment (pgruetter) [reply src/pages/json-schema-viewer.js]: Fixed by @zzeekk in 5ad741e3bab3b1a718afa023e62111eadaaefba3

File: static/json-schema-viewer/schemas/index.json
Comment (pgruetter) [reply static/json-schema-viewer/schemas/index.json:1]: Done in 8dcb7de12b637488f73d486ef8e78d71f29547fa

File: yarn.lock
Comment (pgruetter) [reply yarn.lock:1]: Done in 8dcb7de12b637488f73d486ef8e78d71f29547fa

**Review by pgruetter — CHANGES_REQUESTED** [review]
Looks good overall! Thanks especially for fixing a few formatting problems we apparently had for a long time. 
A few questions:
- I wonder why these escapes where necessary. In my case, they are inserted verbatim.
- Since it requires npm 18 now, do we need to change anything for GitHub pages before we merge this? Github Actions?

File: docs/getting-started/part-1/select-columns.md (hunk 3 above)
Comment (eliasarnold) [reply docs/getting-started/part-1/select-columns.md]: It seems like the code on the repo is not the same as the code deployed to [smartdatalake.ch](https://smartdatalake.ch/docs/getting-started/part-1/select-columns). On the website, this box is of type "DANGER":
image
In the image, the older version of the markdown parser in docusaurus 2 has some problems with the quotation. That is fixed with the new version of MDX (markdown parser).

Comment (pgruetter) [reply docs/getting-started/part-1/select-columns.md]: Ah strange, then you can leave it of course. I think a warning makes more sense here though.

File: blog/2022-04-06-SDL-snowpark/2022-04-06-SDL-snowpark.md (hunk 2 above)
Comment (eliasarnold) [reply blog/2022-04-06-SDL-snowpark/2022-04-06-SDL-snowpark.md]: Thanks for spotting this error! This was a leftover of one of my refactoring attempts.

File: docs/getting-started/part-1/get-departures.md [hunk 4]
```diff
@@ -74,12 +74,13 @@ This data object acts as a target for our first action, so where and how to down
 We set type to JsonFileDataObject because we know from before that the webservice will return a json file.
 Path defines where the file will be stored. You could choose any name you want, but most of the time, the name of your DataObject is a good fit.
 Instead of writing *stg-departures* again,
-we used the placeholder *~{id}* which gets replaced by the DataObject-ID. Don't forget to surround that placeholder
+we used the placeholder *~\{id}* which gets replaced by the DataObject-ID. Don't forget to surround that placeholder
```
Comment (pgruetter) [inline docs/getting-started/part-1/get-departures.md:77]:
```suggestion
we used the placeholder *~{id}* which gets replaced by the DataObject-ID. Don't forget to surround that placeholder
```

File: docs/getting-started/troubleshooting/docker-on-windows.md [hunk 5]
```diff
@@ -30,15 +30,15 @@ For [part 2 of this guide](../part-2/delta-lake-format.md), you need docker-comp
 For composing multiple podman containers, you can just execute our custom script podman-compose.sh from the getting-started base directory.
 
 After running the script in the getting-started folder you should now be able to open Polynote on port localhost:8192, as WSL2 automatically publishes all ports on Windows.
-If the port is not accessible, you can use `wsl hostname -I` on Windows command line to get the IP adress of WSL, and then access Polynote over {ip-address}:8192.
+If the port is not accessible, you can use `wsl hostname -I` on Windows command line to get the IP adress of WSL, and then access Polynote over \{ip-address}:8192.
```
Comment (pgruetter) [inline docs/getting-started/troubleshooting/docker-on-windows.md:33]:
```suggestion
If the port is not accessible, you can use `wsl hostname -I` on Windows command line to get the IP adress of WSL, and then access Polynote over {ip-address}:8192.
```

File: docs/reference/hoconVariables.md [hunk 6]
```diff
@@ -3,13 +3,16 @@ id: hoconVariables
...
+import TabItem from '@theme/TabItem';
+
 # Variables
 Variables can be used to serve various goals. On the one hand, we want to prevent repeating specifications, keep specifications flexible, but also want to be able to change settings at call time. 
 
 Here are some examples listed. 
 
 ## Local Substitution
-Local substitution allows to reuse the id of a configuration object inside its attribute definitions by the special token "~{id}". See the following example:
+Local substitution allows to reuse the id of a configuration object inside its attribute definitions by the special token "~\{id}". See the following example:
```
Comment (pgruetter) [inline docs/reference/hoconVariables.md:15]:
```suggestion
Local substitution allows to reuse the id of a configuration object inside its attribute definitions by the special token "~{id}". See the following example:
```

File: docs/reference/transformations.md [hunk 7]
```diff
@@ -290,7 +293,7 @@ For custom transformers, you can therefore provide additional options:
 
 
 ##### In SQL
-If you want to use options in SQL, the syntax is %{key}:
+If you want to use options in SQL, the syntax is %\{key}:
```
Comment (pgruetter) [inline docs/reference/transformations.md:296]:
```suggestion
If you want to use options in SQL, the syntax is %{key}:
```

**Review by pgruetter — APPROVED** [review]
Still found a few \ escape characters. Other than that, looks good to me now.

**Comment by pgruetter** [comment]
@eliasarnold Thanks, did you see this open question?
- Since it requires npm 18 now, do we need to change anything for GitHub pages before we merge this? Github Actions?

**Comment by eliasarnold** [comment]
I don't think that there are any issues with npm 18, since the dry-run of the deploy (Test deployment check) already uses npm 18:
image

**Comment by eliasarnold** [comment]
I had to revert the last four commits since the escapes are necessary with the new docusaurus version. Now, a word between curly braces (like `{id}`, `{key}`, or `{ip-address}`) is a reserved keyword which has to be escaped.
